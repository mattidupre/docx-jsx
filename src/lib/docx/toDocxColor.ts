import ColorJs from 'colorjs.io';
import type { Color } from '../../entities';

// TODO: Use lighter submodule of ColorJS.

export const toDocxColor = (
  color: Color | undefined | 'currentColor',
): undefined | string => {
  if (!color || color === 'currentColor') {
    return undefined;
  }
  const hex = new ColorJs(color).to('srgb').display({ format: 'hex' });
  if (hex.startsWith('#')) {
    const hexBase = hex.slice(1);
    // convert shorthand f00 to ff0000
    // https://github.com/color-js/color.js/issues/266
    if (hexBase.length === 3) {
      return (
        hexBase[0] +
        hexBase[0] +
        hexBase[1] +
        hexBase[1] +
        hexBase[2] +
        hexBase[2]
      );
    }
    if (hexBase.length === 6) {
      return hexBase;
    }
    if (hexBase.length === 8) {
      console.warn(`Transparency value ignored for color ${color}`);
      return hexBase.slice(0, 6);
    }
  }
  throw new TypeError(`Invalid color ${color}`);
};
