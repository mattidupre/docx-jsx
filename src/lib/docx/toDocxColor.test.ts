import {
  test,
  expect,
  describe,
  vi,
  beforeEach,
  afterEach,
  type SpyInstance,
} from 'vitest';
import { toDocxColor } from './toDocxColor';

let warn: SpyInstance;
beforeEach(() => {
  warn = vi.spyOn(console, 'warn').mockImplementation(() => {});
});

afterEach(() => {
  vi.restoreAllMocks();
});

describe('toDocxColor', () => {
  const expectValue = (value: any) => {
    expect(value).toMatch(/^[0-9a-f]{6}$/);
    expect(value).toMatchSnapshot();
  };

  for (const input of [
    'red',
    '#f00',
    '#ff0000',
    'rgb(255, 0, 0)',
    'oklch(70% 0.197 22.73)',
  ] as const) {
    test(`with ${input}`, () => {
      const result = toDocxColor(input);
      expectValue(result);
    });
  }

  for (const input of ['rgba(255, 0, 0, 0.5)'] as const) {
    test(`with ${input} [WARNS]`, () => {
      const result = toDocxColor(input);
      expectValue(result);
      expect(warn).toHaveBeenCalledTimes(1);
    });
  }

  test('with currentColor', () => {
    expect(toDocxColor('currentColor')).toBe(undefined);
  });

  test('with undefined', () => {
    expect(toDocxColor(undefined)).toBe(undefined);
  });

  test('with ""', () => {
    expect(toDocxColor('')).toBe(undefined);
  });
});
