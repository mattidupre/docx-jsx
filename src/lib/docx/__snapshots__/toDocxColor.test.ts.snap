// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toDocxColor > with #f00 1`] = `"ff0000"`;

exports[`toDocxColor > with #ff0000 1`] = `"ff0000"`;

exports[`toDocxColor > with oklch(70% 0.197 22.73) 1`] = `"ff6062"`;

exports[`toDocxColor > with red 1`] = `"ff0000"`;

exports[`toDocxColor > with rgb(255, 0, 0) 1`] = `"ff0000"`;

exports[`toDocxColor > with rgba(255, 0, 0, 0.5) [WARNS] 1`] = `"ff0000"`;

exports[`toDocxColor > with rgba(255, 0, 0, 0.5) [warns] 1`] = `"ff0000"`;

exports[`toDocxColor > with rgba(255, 0, 0, 0.5) 1`] = `"ff0000"`;
